File: src/web/common/options.c
/**
 * @file options.c
 * @brief WASM shared option setters and getters
 * @ingroup web_common
 *
 * Provides unified option access for all WASM modes.
 * These functions are compiled into both mirror.wasm and client.wasm.
 */

#include "options.h"
#include <ascii-chat/options/options.h>
#include <ascii-chat/options/rcu.h>
#include <ascii-chat/options/parsers.h>
#include <ascii-chat/asciichat_errno.h>
#include <ascii-chat/log/log.h>
#include <stdlib.h>

// ============================================================================
// Dimension Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_width(int value) {
  if (value <= 0 || value > 1000)
    return -1;
  asciichat_error_t err = options_set_int("width", value);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_width(void) {
  return GET_OPTION(width);
}

EMSCRIPTEN_KEEPALIVE
int set_height(int value) {
  if (value <= 0 || value > 1000)
    return -1;
  asciichat_error_t err = options_set_int("height", value);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_height(void) {
  return GET_OPTION(height);
}

// ============================================================================
// Color Mode Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_color_mode(int mode) {
  asciichat_error_t err = options_set_int("color_mode", mode);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_color_mode(void) {
  return GET_OPTION(color_mode);
}

// ============================================================================
// Color Filter Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_color_filter(int filter) {
  asciichat_error_t err = options_set_int("color_filter", filter);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_color_filter(void) {
  return GET_OPTION(color_filter);
}

// ============================================================================
// Palette Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_palette(const char *palette_name) {
  if (!palette_name)
    return -1;

  palette_type_t palette_value;
  char *error_msg = NULL;

  if (!parse_palette_type(palette_name, &palette_value, &error_msg)) {
    if (error_msg) {
      log_error("Failed to parse palette '%s': %s", palette_name, error_msg);
      free(error_msg);
    }
    return -1;
  }

  asciichat_error_t err = options_set_int("palette_type", (int)palette_value);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_palette(void) {
  return GET_OPTION(palette_type);
}

EMSCRIPTEN_KEEPALIVE
int set_palette_chars(const char *chars) {
  if (!chars)
    return -1;
  asciichat_error_t err = options_set_string("palette_custom", chars);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
const char *get_palette_chars(void) {
  return GET_OPTION(palette_custom);
}

// ============================================================================
// Matrix Rain Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_matrix_rain(int enabled) {
  asciichat_error_t err = options_set_bool("matrix_rain", enabled != 0);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_matrix_rain(void) {
  return GET_OPTION(matrix_rain) ? 1 : 0;
}

// ============================================================================
// Horizontal Flip Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_flip_x(int enabled) {
  asciichat_error_t err = options_set_bool("flip_x", enabled != 0);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_flip_x(void) {
  return GET_OPTION(flip_x) ? 1 : 0;
}

// ============================================================================
// Render Mode Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_render_mode(int mode) {
  asciichat_error_t err = options_set_int("render_mode", mode);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_render_mode(void) {
  return GET_OPTION(render_mode);
}

// ============================================================================
// Target FPS Accessors
// ============================================================================

EMSCRIPTEN_KEEPALIVE
int set_target_fps(int fps) {
  if (fps < 15 || fps > 60)
    return -1;
  asciichat_error_t err = options_set_int("fps", fps);
  return (err == ASCIICHAT_OK) ? 0 : -1;
}

EMSCRIPTEN_KEEPALIVE
int get_target_fps(void) {
  return GET_OPTION(fps);
}

// ============================================================================
// Help Text API
// ============================================================================

/**
 * Get help text for a CLI option in a specific mode
 * Exported to WASM for JavaScript access via FFI
 *
 * @param mode The mode (asciichat_mode_t as int)
 * @param option_name The long name of the option
 * @return Help text string or NULL if not applicable
 */
EMSCRIPTEN_KEEPALIVE
const char *get_help_text(int mode, const char *option_name) {
  if (!option_name || !option_name[0])
    return NULL;
  asciichat_mode_t mode_enum = (asciichat_mode_t)mode;
  return options_get_help_text(mode_enum, option_name);
}

File: web/packages/shared/src/wasm/common/options.ts
/**
 * @file options.ts
 * @brief Unified WASM options API for mirror and client modes
 *
 * Provides a single, mode-agnostic API for all WASM option accessors.
 * Both mirror.ts and client.ts initialize this module with their own
 * options accessor, and React components use these functions directly.
 *
 * This eliminates code duplication between mirror and client implementations.
 */

import type { OptionAccessor } from "./optionsWrapper";

// Enums matching libasciichat definitions
export enum RenderMode {
  FOREGROUND = 0,
  BACKGROUND = 1,
  HALF_BLOCK = 2,
}

export enum ColorMode {
  AUTO = -1,
  NONE = 0,
  COLOR_16 = 1,
  COLOR_256 = 2,
  TRUECOLOR = 3,
}

export enum ColorFilter {
  NONE = 0,
  BLACK = 1,
  WHITE = 2,
  GREEN = 3,
  MAGENTA = 4,
  FUCHSIA = 5,
  ORANGE = 6,
  TEAL = 7,
  CYAN = 8,
  PINK = 9,
  RED = 10,
  YELLOW = 11,
  RAINBOW = 12,
}

export type Palette =
  | "standard"
  | "blocks"
  | "digital"
  | "minimal"
  | "cool"
  | "custom";

// Module-level options accessor (set by mirror.ts or client.ts during init)
let options: OptionAccessor | null = null;

// TODO: Query C code's options struct directly instead of caching in JavaScript.
// Currently caching palette, flipX, flipY in JavaScript because WASM getters
// (_get_palette, _get_flip_x, _get_flip_y) return unreliable values during rendering.
// The real fix: the C code has an internal options struct with the actual state.
// Instead of JavaScript caching, we should expose a getter that queries that struct
// directly, making the C code the single source of truth. This would eliminate
// the need for JavaScript-side caching and ensure mirror_init_with_args works
// correctly with the real state values.
//
// Known bugs (WASM only, don't appear in terminal):
// 1. --palette passed to mirror_init_with_args returns different values each time
// 2. --flip-x didn't work correctly (needs investigation, may require querying C struct first)
// 3. --matrix doesn't render green on first frame after resize (renders truecolor instead)
// These issues don't occur in the terminal version, suggesting WASM-specific state handling.

// Cache for values since WASM getters may not track set values reliably
let cachedPalette: Palette = "standard";
let cachedPaletteChars: string = "";
let cachedFlipX: boolean = false;
let cachedFlipY: boolean = false;

/**
 * Initialize the options module with an options accessor
 * Called by mirror.ts or client.ts after WASM module initialization
 * @internal
 */
export function initializeOptions(accessor: OptionAccessor): void {
  options = accessor;
  cachedPalette = "standard";
  cachedPaletteChars = "";
  cachedFlipX = false;
  cachedFlipY = false;
}

/**
 * Cleanup options module
 * @internal
 */
export function cleanupOptions(): void {
  options = null;
  cachedPalette = "standard";
  cachedPaletteChars = "";
  cachedFlipX = false;
  cachedFlipY = false;
}

/**
 * Check if the options module has been initialized
 * @returns true if options are ready, false otherwise
 */
export function isOptionsInitialized(): boolean {
  return options !== null;
}

// ============================================================================
// Dimension Accessors
// ============================================================================

export function setWidth(width: number): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("width", width);
}

export function getWidth(): number {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("width");
}

export function setHeight(height: number): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("height", height);
}

export function getHeight(): number {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("height");
}

export function setDimensions(width: number, height: number): void {
  setWidth(width);
  setHeight(height);
}

export function getDimensions(): { width: number; height: number } {
  return {
    width: getWidth(),
    height: getHeight(),
  };
}

// ============================================================================
// Color Mode Accessors
// ============================================================================

export function setColorMode(mode: ColorMode): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("color_mode", mode);
}

export function getColorMode(): ColorMode {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("color_mode");
}

// ============================================================================
// Color Filter Accessors
// ============================================================================

export function setColorFilter(filter: ColorFilter): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("color_filter", filter);
}

export function getColorFilter(): ColorFilter {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("color_filter");
}

// ============================================================================
// Render Mode Accessors
// ============================================================================

export function setRenderMode(mode: RenderMode): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("render_mode", mode);
}

export function getRenderMode(): RenderMode {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("render_mode");
}

// ============================================================================
// Palette Accessors
// ============================================================================

export function setPalette(palette: Palette): void {
  if (!options) throw new Error("Options not initialized");
  options.setString("palette", palette);
  cachedPalette = palette;
}

export function getPalette(): Palette {
  if (!options) throw new Error("Options not initialized");
  return cachedPalette;
}

export function setPaletteChars(chars: string): void {
  if (!options) throw new Error("Options not initialized");
  options.setString("palette_chars", chars);
  cachedPaletteChars = chars;
}

export function getPaletteChars(): string {
  if (!options) throw new Error("Options not initialized");
  return cachedPaletteChars;
}

// ============================================================================
// Matrix Rain Accessors
// ============================================================================

export function setMatrixRain(enabled: boolean): void {
  if (!options) throw new Error("Options not initialized");
  options.setBool("matrix_rain", enabled);
}

export function getMatrixRain(): boolean {
  if (!options) throw new Error("Options not initialized");
  return options.getBool("matrix_rain");
}

// ============================================================================
// Horizontal Flip Accessors
// ============================================================================

export function setFlipX(enabled: boolean): void {
  if (!options) throw new Error("Options not initialized");
  options.setBool("flip_x", enabled);
  cachedFlipX = enabled;
}

export function getFlipX(): boolean {
  if (!options) throw new Error("Options not initialized");
  return cachedFlipX;
}

// ============================================================================
// Vertical Flip Accessors
// ============================================================================

export function setFlipY(enabled: boolean): void {
  if (!options) throw new Error("Options not initialized");
  options.setBool("flip_y", enabled);
  cachedFlipY = enabled;
}

export function getFlipY(): boolean {
  if (!options) throw new Error("Options not initialized");
  return cachedFlipY;
}

// ============================================================================
// Target FPS Accessors
// ============================================================================

export function setTargetFps(fps: number): void {
  if (!options) throw new Error("Options not initialized");
  options.setInt("target_fps", fps);
}

export function getTargetFps(): number {
  if (!options) throw new Error("Options not initialized");
  return options.getInt("target_fps");
}
